### PR TITLE
Zachmu/analyzer

### PR DESCRIPTION
**Pull request by zachmu**
Analyzer improvements, mostly related to treatment of aliases to support selecting the same table twice, but lots of other improvements as well.

### CODE REVIEW

**Comment by zachmu** [comment]
Pretty well tested, but take a quick look for any obvious problems.

File: sql/analyzer/analyzer.go [hunk 1]
```diff
@@ -136,7 +144,11 @@ func (ab *Builder) Build() *Analyzer {
 // Analyzer analyzes nodes of the execution plan and applies rules and validations
 // to them.
 type Analyzer struct {
+	// Whether to log various debugging messages
 	Debug       bool
+	// Whether to output the query plan at each step of the analyzer
+	Verbose     bool
+	debugCtx    []string
```
Comment (bheni) [inline sql/analyzer/analyzer.go:151]: Reading the code below I had to scroll back up to see what debugCtx was.  I read Ctx and think context.Context.  Feel like another name would clear up some confusion.

File: sql/analyzer/fix_field_indexes.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		var err error
+		result[i], err = fixFieldIndexes(schema, e)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+// fixFieldIndexes transforms the given expression setting correct indexes
```
Comment (bheni) [inline sql/analyzer/fix_field_indexes.go:22]: "the given expression setting correct indexes for GetField"

 wat?

File: sql/analyzer/resolve_columns_test.go [hunk 3]
```diff
@@ -76,7 +77,7 @@ func TestMisusedAlias(t *testing.T) {
 }
 
 func TestQualifyColumns(t *testing.T) {
-	require := require.New(t)
```
[the remark this answers is not in the document]
Comment (zachmu) [reply sql/analyzer/resolve_columns_test.go:79]: require stops the test on the first failed check, assert lets it continue.

File: sql/analyzer/resolve_columns.go [hunk 4]
```diff
@@ -334,8 +339,19 @@ func findChildIndexedColumns(n sql.Node) map[tableCol]indexedCol {
 	var idx int
 	var columns = make(map[tableCol]indexedCol)
 
+	aliases := getTableAliases(n)
+
 	for _, child := range n.Children() {
-		for _, col := range child.Schema() {
+		childSch := child.Schema()
+		for _, col := range childSch {
+			// columns of tables with an alias can be referred to either by their aliased or unaliased name, so add an entry
```
[the remark this answers is not in the document]
Comment (zachmu) [reply sql/analyzer/resolve_columns.go:347]: Done.

File: sql/analyzer/analyzer.go (hunk 1 above)
Comment (zachmu) [reply sql/analyzer/analyzer.go:151]: Done

File: sql/analyzer/fix_field_indexes.go (hunk 2 above)
Comment (zachmu) [reply sql/analyzer/fix_field_indexes.go:22]: Done

**Pull request merged**